src/ir/mod.rs

```diff
@@ -391,6 +397,9 @@ impl<B: Backend> Compiler<B> {
 
         Ok(())
     }
+    fn resolve_location(&self, loc: Location) -> &str {
+        self.files.source_slice(loc.file, loc.span).unwrap() // TODO unwrap safely
```

Comment:
I just had an idea: if we made the fields on Location private we could ensure all locations came from the lexer, and so are always in bounds (modulo bugs).